src/tri.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright 2014-2024 bluss and ndarray developers.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use core::cmp::{max, min};
+
+use num_traits::Zero;
+
+use crate::{Array, ArrayBase, Axis, Data, Dimension, IntoDimension, RemoveAxis, Zip};
+
+impl<S, A, D> ArrayBase<S, D>
+where
+    S: Data<Elem = A>,
+    D: Dimension + RemoveAxis,
+    A: Clone + Zero,
+    D::Smaller: Copy,
+{
+    /// Upper triangular of an array.
+    ///
+    /// Return a copy of the array with elements below the *k*-th diagonal zeroed.
+    /// For arrays with `ndim` exceeding 2, `triu` will apply to the final two axes.
+    ///
+    /// See also [`ArrayBase::tril`]
+    ///
+    /// ```
+    /// use ndarray::array;
+    ///
+    /// let arr = array![[1, 2, 3], [4, 5, 6], [7, 8, 9]];
+    /// let res = arr.triu(0);
+    /// assert_eq!(res, array![[1, 2, 3], [0, 5, 6], [0, 0, 9]]);
+    /// ```
+    pub fn triu(&self, k: isize) -> Array<A, D> {
+        let mut res = Array::zeros(self.raw_dim());
+        Zip::indexed(self.rows())
+            .and(res.rows_mut())
+            .for_each(|i, src, mut dst| {
+                let row_num = i.into_dimension().last_elem();
+                let lower: isize = max(row_num as isize + k, 0);
+                dst.slice_mut(s![lower..]).assign(&src.slice(s![lower..]));
+            });
+
+        res
+    }
+
+    /// Lower triangular of an array.
+    ///
+    /// Return a copy of the array with elements above the *k*-th diagonal zeroed.
+    /// For arrays with `ndim` exceeding 2, `tril` will apply to the final two axes.
+    ///
+    /// See also [`ArrayBase::triu`]
+    ///
+    /// ```
+    /// use ndarray::array;
+    ///
+    /// let arr = array![[1, 2, 3], [4, 5, 6], [7, 8, 9]];
+    /// let res = arr.tril(0);
+    /// assert_eq!(res, array![[1, 0, 0], [4, 5, 0], [7, 8, 9]]);
+    /// ```
+    pub fn tril(&self, k: isize) -> Array<A, D> {
+        let mut res = Array::zeros(self.raw_dim());
+        let ncols = self.len_of(Axis(self.ndim() - 1)) as isize;
+        Zip::indexed(self.rows())
+            .and(res.rows_mut())
+            .for_each(|i, src, mut dst| {
+                let row_num = i.into_dimension().last_elem();
+                let upper: isize = min(row_num as isize + k, ncols) + 1;
+                dst.slice_mut(s![..upper]).assign(&src.slice(s![..upper]));
+            });
+
+        res
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::array;
+
+    #[test]
+    fn test_2d() {
```

Comment:
So, what happens on 1D array?

Reply:
Excellent question, and I think it comes down to design - NumPy broadcasts the 1D array into a square matrix, then does the triangular, like this:
```python
>>> np.triu([1, 2, 3])
array([[1, 2, 3],
       [0, 2, 3],
       [0, 0, 3]])
```
However, it seems from their code that this is an implementation result rather than a design decision? Since they use a mask rather than my direct iteration, causing broadcasting.

I think we have three options:
1. Copy NumPy
2. Use a `Result`
3. If possible, use some sort of trait bound, or only implement it for arrays that are 2D or more.
I like (3) the most, but would love some input.

Reply:
Ok, the behavior as it is here just... doesn't change 1D arrays. And I think that's a reasonable result that I hadn't really thought of.

Reply:
I just tested numpy's behavior and... wow, it transform the 1D array into a 2D array, something that we can't do in ndarray. I think what you did is great.